src/main/java/com/lineate/testyourlexicon/services/GameService.java

```diff
@@ -47,13 +54,116 @@ public SupportedLanguagesDto supportedLanguages() {
+    String translateFrom = gameConfiguration.getTranslateFrom();
+    String translateTo = gameConfiguration.getTranslateTo();
+    int answerOptionsCount = gameConfiguration.getAnswerCount();
+    Question question = null;
+    do {
+      question = translationService.getRandomQuestion(translateFrom,
+        translateTo, answerOptionsCount);
+    } while (questionRepository.existsByGameAndTranslationId(game, question.getTranslationId()));
+    return question;
+  }
+
+  public QuestionEntity saveQuestionInDatabase(Question question, Game game) {
+    QuestionEntity questionEntity = new QuestionEntity();
+    questionEntity.setGuessed(false);
+    questionEntity.setTranslationId(question.getTranslationId());
+    questionEntity.setGame(game);
+    return questionRepository.save(questionEntity);
+  }
+
+  public Game updateGameCurrentQuestion(Game game, QuestionEntity questionEntity) {
+    game.setStepsLeft(game.getStepsLeft() - 1);
+    game.setCurrentQuestionId(questionEntity.getId());
+    return gameRepository.save(game);
+  }
+
+  public StepDto userActiveGameNextStep(User user, long gameId) {
+    Game game = validateGameForUser(user, gameId);
+    Question question = generateRandomQuestion(user.getGameConfiguration(), game);
+    QuestionEntity questionEntity = saveQuestionInDatabase(question, game);
+    game = updateGameCurrentQuestion(game, questionEntity);
+    log.info("Generated step {'user': {}, 'game_id': {}, 'question_id': {}}",
+        user.getId(), game.getGameId(), questionEntity.getId());
+    return new StepDto(question, gameId);
+  }
+
+  public QuestionEntity getQuestion(Long questionId) {
+    if (questionId == null) {
+      throw new GeneralMessageException("Couldn't not find active question");
+    }
+    Optional<QuestionEntity> questionEntity =
+      questionRepository.findById(questionId);
+    if (questionEntity.isEmpty()) {
+      throw new GeneralMessageException("Invalid question id");
+    }
+    return questionEntity.get();
+  }
+
+  public AnswerResponseDto userActiveGameAnswer(User user,
+                                   AnswerRequestDto answerRequestDto,
+                                   long gameId) {
+    Game game = validateGameForUser(user, gameId);
+    QuestionEntity questionEntity = getQuestion(game.getCurrentQuestionId());
+
+    // Get correct answer
+    Long translationId = questionEntity.getTranslationId();
+    String language = user.getGameConfiguration().getTranslateTo();
+    String correctAnswer =
+         translationRepository.languageDefinitionGivenIdAndLanguage(translationId, language);
+
+    // Check if users answer is the same as correct one
+    boolean guessed = correctAnswer.equalsIgnoreCase(answerRequestDto.getAnswer());
+    if (guessed) {
+      questionEntity.setGuessed(true);
+      questionRepository.save(questionEntity);
+    }
+    game.setCurrentQuestionId(null);
+    gameRepository.save(game);
```

Comment:
What will happen if for some reason app will fail to complete this operation (suppose that DB becomes unavailable). There might be an inconsistent state in the database since the question was added, but the game was not yet updated. Can we protect ourselves from that? Hint: [@Transactional](https://www.baeldung.com/transaction-configuration-with-jpa-and-spring)